File: src/main/kotlin/com/tft/apibatch/mongo/MongoConfig.java
package com.tft.apibatch.mongo;

import com.mongodb.*;
import com.mongodb.connection.ConnectionPoolSettings.Builder;

import java.util.concurrent.TimeUnit;

//@Configuration
public interface MongoConfig {

    default MongoClientSettings mongoClientSettings(String uri, Block<Builder> poolSettings) {

        ConnectionString connectionString = new ConnectionString(uri);

        return MongoClientSettings.builder()
                .retryReads(false)
                .retryWrites(false)
                .writeConcern(WriteConcern.ACKNOWLEDGED)
                .readPreference(ReadPreference.secondaryPreferred())
                .applyConnectionString(connectionString)
                .applyToConnectionPoolSettings(poolSettings)
                .applyToSocketSettings(x -> {
                    x.connectTimeout(3, TimeUnit.SECONDS);
                    x.readTimeout(2, TimeUnit.SECONDS);
                })
                .build();
    }

    default MongoClientSettings mongoClientSettings(String uri, String userName, String dbname, String password, Block<Builder> poolSettings) {

        ConnectionString connectionString = new ConnectionString(uri);

        return MongoClientSettings.builder()
                .credential(MongoCredential.createCredential(userName, dbname, password.toCharArray()))
                .retryReads(false)
                .retryWrites(false)
                .writeConcern(WriteConcern.ACKNOWLEDGED)
                .readPreference(ReadPreference.secondaryPreferred())
                .applyConnectionString(connectionString)
                .applyToConnectionPoolSettings(poolSettings)
                .applyToSocketSettings(x -> {
                    x.connectTimeout(3, TimeUnit.SECONDS);
                    x.readTimeout(2, TimeUnit.SECONDS);
                })
                .build();
    }

    default Block<Builder> generalPoolSettings() {
        return x -> {
            x.maxSize(50);
            x.maxWaitTime(3, TimeUnit.SECONDS);
        };
    }

}

File: src/main/kotlin/com/tft/apibatch/mongo/MongoConfig.kt
package com.tft.apibatch.mongo

import com.mongodb.*
import com.mongodb.connection.ConnectionPoolSettings
import com.mongodb.connection.SocketSettings
import java.util.concurrent.TimeUnit

//@Configuration
interface MongoConfig {
    fun mongoClientSettings(uri: String?, poolSettings: Block<ConnectionPoolSettings.Builder>): MongoClientSettings {
        val connectionString = ConnectionString(uri!!)
        return MongoClientSettings.builder()
            .retryReads(false)
            .retryWrites(false)
            .writeConcern(WriteConcern.ACKNOWLEDGED)
            .readPreference(ReadPreference.secondaryPreferred())
            .applyConnectionString(connectionString)
            .applyToConnectionPoolSettings(poolSettings!!)
            .applyToSocketSettings { x: SocketSettings.Builder ->
                x.connectTimeout(3, TimeUnit.SECONDS)
                x.readTimeout(2, TimeUnit.SECONDS)
            }
            .build()
    }

    fun mongoClientSettings(
        uri: String?,
        userName: String?,
        dbname: String?,
        password: String,
        poolSettings: Block<ConnectionPoolSettings.Builder?>?
    ): MongoClientSettings? {
        val connectionString = ConnectionString(uri!!)
        return MongoClientSettings.builder()
            .credential(MongoCredential.createCredential(userName!!, dbname!!, password.toCharArray()))
            .retryReads(false)
            .retryWrites(false)
            .writeConcern(WriteConcern.ACKNOWLEDGED)
            .readPreference(ReadPreference.secondaryPreferred())
            .applyConnectionString(connectionString)
            .applyToConnectionPoolSettings(poolSettings!!)
            .applyToSocketSettings { x: SocketSettings.Builder ->
                x.connectTimeout(3, TimeUnit.SECONDS)
                x.readTimeout(2, TimeUnit.SECONDS)
            }
            .build()
    }

    fun generalPoolSettings(): Block<ConnectionPoolSettings.Builder> {
        return Block { x: ConnectionPoolSettings.Builder ->
            x.maxSize(50)
            x.maxWaitTime(3, TimeUnit.SECONDS)
        }
    }
}